### PR TITLE
BBR complete code

### PR DESCRIPTION
**Pull request by fedech**
Fixed issues in Umberto's PR (#6 BBR-Related Fixes), merging with scheduling PR (#5 Scheduling) and bugfixes (#7 Update quic-bbr.cc), fixed formatting according to ns-3 style guide, fixed connection closing bug.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,12 +1,18 @@
-QUIC implementation for ns-3
+QUIC implementation in ns-3
 ================================
 
 ## QUIC code base
-This repository contains in the code for a native IETF QUIC implementation in ns-3.
+
+This repository is a fork of [signetlabdei/quic](https://github.com/signetlabdei/quic) which contains in the code for a native IETF QUIC implementation in ns-3.
```
Comment (mychele) [inline README.md:6]: check

File: README.md [hunk 2]
```diff
@@ -18,16 +24,12 @@ Required dependencies include git and a build environment.
 
 #### Installing dependencies ####
 
-Please refer to [the ns-3 wiki](https://www.nsnam.org/wiki/Installation) for instructions on how to set up your system to install ns-3.
+Please refer to [the ns-3 wiki](https://www.nsnam.org/wiki/Installation) for instructions on how to set up your system to install ns-3. 
+This module was implemented and tested on ns-3.29, and forward compatibility is not guaranteed.
```
Comment (mychele) [inline README.md:28]: check and update

File: model/quic-bbr.cc [hunk 3]
```diff
@@ -549,16 +551,17 @@ QuicBbr::WhichState (BbrMode_t mode) const
...
-    case 3:
-      return "BBR_PROBE_RTT";
+      case 0:
+        return "BBR_STARTUP";
+      case 1:
+        return "BBR_DRAIN";
+      case 2:
+        return "BBR_PROBE_BW";
+      case 3:
+        return "BBR_PROBE_RTT";
```
Comment (mychele) [inline model/quic-bbr.cc:561]: The correct way to handle this is to add a default case to the switch, which triggers the error

File: model/quic-congestion-ops.cc [hunk 4]
```diff
@@ -44,8 +45,8 @@ TypeId
 QuicCongestionOps::GetTypeId (void)
 {
   static TypeId tid = TypeId ("ns3::QuicCongestionControl").SetParent<
-      TcpNewReno> ().SetGroupName ("Internet").AddConstructor<
-      QuicCongestionOps> ();
+    TcpNewReno> ().SetGroupName ("Internet").AddConstructor<
```
Comment (mychele) [inline model/quic-congestion-ops.cc:48]: check indent

File: model/quic-socket-base.cc [hunk 5]
```diff
@@ -115,50 +117,50 @@ QuicSocketBase::GetTypeId (void)
                    MakeUintegerAccessor (&QuicSocketBase::m_initial_max_stream_id_bidi),
                    MakeUintegerChecker<uint32_t> ())
     .AddAttribute (
```
Comment (mychele) [inline model/quic-socket-base.cc:119]: check indent of the attributes (maybe it's just the github interface)

File: model/quic-socket-base.cc [hunk 6]
```diff
@@ -296,6 +308,16 @@ QuicSocketBase::GetTypeId (void)
                      "Receive QUIC packet from UDP protocol",
                      MakeTraceSourceAccessor (&QuicSocketBase::m_rxTrace),
                      "ns3::QuicSocketBase::QuicTxRxTracedCallback")
+    .AddAttribute ("CongControl",
+                   "The Object managing CC operations",
+                   PointerValue (),
```
Comment (mychele) [inline model/quic-socket-base.cc:313]: Are you sure a default is not needed?

File: model/quic-socket-base.cc [hunk 7]
```diff
@@ -577,6 +602,8 @@ QuicSocketBase::QuicSocketBase (const QuicSocketBase& sock)   // Copy constructo
 {
   NS_LOG_FUNCTION (this);
   NS_LOG_LOGIC (this << " invoked the copy constructor");
+  m_lastMaxData = 0;
```
Comment (mychele) [inline model/quic-socket-base.cc:605]: These should go in the member initialization list

File: model/quic-socket-base.cc [hunk 8]
```diff
@@ -764,7 +791,8 @@ QuicSocketBase::Listen (void)
       return -1;
     }
 
-  NS_ASSERT (m_quicl4->SetListener (this));
+  bool res = m_quicl4->SetListener (this);
+  NS_ASSERT (res);
```
Comment (mychele) [inline model/quic-socket-base.cc:795]:
```suggestion
  NS_ASSERT_MSG (res, "Failed to set the listener");
```

File: model/quic-socket-base.cc [hunk 9]
```diff
@@ -1907,6 +1937,16 @@ QuicSocketBase::SetConnectionId (uint64_t connectionId)
   m_connectionId = connectionId;
 }
 
+void
+QuicSocketBase::InitializeScheduling ()
```
Comment (mychele) [inline model/quic-socket-base.cc:1941]: There was a question on this in the previous review - can you check?

"Why are you calling this explicitly from L4 and not in the constructor for QuicSocketBase?"

File: model/quic-socket-base.cc [hunk 10]
```diff
@@ -2192,22 +2232,33 @@ QuicSocketBase::OnSendingAckFrame ()
...
-      gaps, additionalAckBlocks);
-  QuicSubheader maxData = QuicSubheader::CreateMaxData(m_quicl5->GetMaxData());
+    largestAcknowledged.GetValue (), ack_delay, largestAcknowledged.GetValue (),
+    gaps, additionalAckBlocks);
 
   Ptr<Packet> ackFrame = Create<Packet> ();
   ackFrame->AddHeader (sub);
-  ackFrame->AddHeader(maxData);
+
+  if (m_lastMaxData < m_maxDataInterval)
```
Comment (mychele) [inline model/quic-socket-base.cc:2251]: Does this solve https://github.com/signetlabdei/quic-ns-3/issues/6?

File: model/quic-socket-tx-buffer.cc [hunk 11]
```diff
@@ -34,46 +34,43 @@
...
-      other.m_lost),
-    m_retrans (other.m_retrans),
-    m_sacked (other.m_sacked),
-    m_acked (other.m_acked),
-    m_isStream (
-      other.m_isStream),
-    m_isStream0 (other.m_isStream0),
-    m_lastSent (
-      other.m_lastSent)
+QuicSocketTxItem::QuicSocketTxItem () :
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:57]: check indent (maybe one for each line is better)

Comment (mychele) [reply model/quic-socket-tx-buffer.cc:57]: also below

File: model/quic-socket-tx-buffer.cc [hunk 12]
```diff
@@ -93,14 +90,88 @@ QuicSocketTxItem::Print (std::ostream &os) const
...
 NS_OBJECT_ENSURE_REGISTERED (QuicSocketTxBuffer);
 
-TypeId
-QuicSocketTxBuffer::GetTypeId (void)
+TypeId QuicSocketTxBuffer::GetTypeId (void)
 {
   static TypeId tid =
-    TypeId ("ns3::QuicSocketTxBuffer").SetParent<Object> ().SetGroupName (
-      "Internet").AddConstructor<QuicSocketTxBuffer> ()
+    TypeId ("ns3::QuicSocketTxBuffer").SetParent<Object>().SetGroupName (
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:173]: check indent

File: model/quic-socket-tx-buffer.cc [hunk 13]
```diff
@@ -209,44 +279,39 @@ QuicSocketTxBuffer::Add (Ptr<Packet> p)
...
+      outItem->m_lastSent = Now ();
+      outItem->m_packet = currentPacket;
+      outItem->m_isStream0 = (*it)->m_isStream0;
+      m_streamZeroList.erase (it);
+      m_streamZeroSize -= currentPacket->GetSize ();
+      m_sentList.insert (m_sentList.end (), outItem);
+      m_sentSize += outItem->m_packet->GetSize ();
+      --m_numFrameStream0InBuffer;
+      Ptr<Packet> toRet = outItem->m_packet->Copy ();
+      return toRet;
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:303]: why do you have a return before it++?

File: model/quic-socket-tx-buffer.cc [hunk 14]
```diff
@@ -209,44 +279,39 @@ QuicSocketTxBuffer::Add (Ptr<Packet> p)
...
       it++;
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:304]: In general, with iterators, use ++it

File: model/quic-socket-tx-buffer.cc [hunk 15]
```diff
@@ -209,44 +279,39 @@ QuicSocketTxBuffer::Add (Ptr<Packet> p)
...
 
-Ptr<Packet>
-QuicSocketTxBuffer::NextStream0Sequence (const SequenceNumber32 seq)
+Ptr<Packet> QuicSocketTxBuffer::NextStream0Sequence (
+  const SequenceNumber32 seq)
 {
   NS_LOG_FUNCTION (this << seq);
 
-  QuicSocketTxItem* outItem = new QuicSocketTxItem ();
+  QuicSocketTxItem *outItem = new QuicSocketTxItem ();
```
Comment (mychele) [inline model/quic-socket-tx-buffer.cc:287]: Why do you still have raw pointers?

File: model/quic-socket-tx-edf-scheduler.h [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+public:
+  /**
+   * \brief Get the type ID.
+   * \return the object TypeId
+   */
+  static TypeId GetTypeId (void);
+
+  QuicSocketTxEdfScheduler ();
+  QuicSocketTxEdfScheduler (const QuicSocketTxEdfScheduler &other);
+  ~QuicSocketTxEdfScheduler (void);
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.h:51]:
```suggestion
  virtual ~QuicSocketTxEdfScheduler (void);
```

File: model/quic-socket-tx-edf-scheduler.h [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+
+  /**
+   * Print the scheduler buffer information to a string
+   *
+   * \param os the std::ostream object
+   */
+  void Print (std::ostream &os) const;
+
+  void Add (Ptr<QuicSocketTxItem> item, bool retx);
```
Comment (mychele) [inline model/quic-socket-tx-edf-scheduler.h:60]:
```suggestion
  void Add (Ptr<QuicSocketTxItem> item, bool retx) override;
```

File: model/quic-socket-tx-pfifo-scheduler.h [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+public:
+  /**
+   * \brief Get the type ID.
+   * \return the object TypeId
+   */
+  static TypeId GetTypeId (void);
+
+  QuicSocketTxPFifoScheduler ();
+  QuicSocketTxPFifoScheduler (const QuicSocketTxPFifoScheduler &other);
+  ~QuicSocketTxPFifoScheduler (void);
```
Comment (mychele) [inline model/quic-socket-tx-pfifo-scheduler.h:50]:
```suggestion
  virtual ~QuicSocketTxPFifoScheduler (void);
```

File: model/quic-socket-tx-pfifo-scheduler.h [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+  /**
+   * Print the scheduler buffer information to a string
+   *
+   * \param os the std::ostream object
+   */
+  void Print (std::ostream &os) const;
+
+  void Add (Ptr<QuicSocketTxItem> item, bool retx);
```
Comment (mychele) [inline model/quic-socket-tx-pfifo-scheduler.h:59]: override

File: model/quic-socket-tx-scheduler.h [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+
+class QuicSocketTxItem;
+
+/**
+ * \ingroup quic
+ *
+ * \brief Tx socket buffer for QUIC
+ */
+class QuicSocketTxScheduler : public Object
+{
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:40]: declare a virtual destructor

File: model/quic-socket-tx-scheduler.h [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+  ~QuicSocketTxFifoScheduler (void);
+
+  /**
+   * Print the scheduler buffer information to a string
+   *
+   * \param os the std::ostream object
+   */
+  void Print (std::ostream &os) const;
+
+  void Add (Ptr<QuicSocketTxItem> item, bool retx);
```
Comment (mychele) [inline model/quic-socket-tx-scheduler.h:108]: override

**Review by mychele — CHANGES_REQUESTED** [review]
As I was mentioning in #5 I do not love the code duplication of the schedulers - but I guess it's too much of an effort

File: README.md (hunk 2 above)
Comment (fedech) [reply README.md:28]: Updated to add ns-3.30

**Comment by mychele** [comment]
There are a couple of other issues (besides the review comments) that I would like to address before this and #9 are merged. Currently, this PR does not compile on macOS with clang, because `virtual void CongControl (Ptr<QuicSocketState> tcb, const struct RateSample *rs);
` in `QuicBbr` hides a virtual function in the base class `TcpCongestionOps`. @fedech is this done on purpose, or is it an error that we bring forward from the early merge of #3 ?

**Comment by fedech** [comment]
> There are a couple of other issues (besides the review comments) that I would like to address before this and #9 are merged. Currently, this PR does not compile on macOS with clang, because `virtual void CongControl (Ptr<QuicSocketState> tcb, const struct RateSample *rs); ` in `QuicBbr` hides a virtual function in the base class `TcpCongestionOps`. @fedech is this done on purpose, or is it an error that we bring forward from the early merge of #3 ?

This was changed in a later version of ns-3 that included those functions in TcpCongestionOps (I was working on 3.29, which didn't have it). I'm switching to ns-3.32 and testing everything before updating the PR

File: model/quic-socket-tx-scheduler.h (hunk 21 above)
Comment (fedech) [reply model/quic-socket-tx-scheduler.h:108]: This does not override anything, QuicSocketTxScheduler doesn't inherit from anyone

File: model/quic-socket-base.cc (hunk 6 above)
Comment (fedech) [reply model/quic-socket-base.cc:313]: The attribute was actually unused, it was probably a residue from some tests. Removed

**Pull request closed without merging**